### ISSUE TITLE
`string` type values incorrectly accept `string` library object

### ISSUE THREAD

**Issue by justarandomgeek** [EmmyLuaLs/emmylua-analyzer-rust#1165]
```lua
---@type string
local s1 = {}  -- this correctly says that table cannot be assigned to string

---@type string
local s2 = string -- this, however, allows the table `string` to be assigned to a string, which should emit the same warning
```

the `string` library object should probably be a type like `std.stringlib` with the `string` value type inheriting from it, so that the `string` type gets the members but does not accept the library object

**Comment by justarandomgeek** [comment]
this is becoming a real nightmare because it also cascades into unions of `string` and other interface types trying to match tables against stringlib instead

**Comment by CppCXY** [comment]
It sounds serious the way you put it, but I don't understand what's happening at all, and we won't choose to change that in the short term either.
The string table itself, as a string type, is a convention that has long been widely adopted. It benefits the extension of the standard library (and this is also the common practice), so even with sufficient reason, I would not choose to change that.

**Comment by justarandomgeek** [comment]
okay, i banged on it some to try to narrow down a specific example where it goes wrong (i've been fighting with some hairy generics trying to get `on_event()` to do type mapping magic from a class assembled of partials and it keeps coming up there, but it's always too much going on to make a good example)

```
---@generic E : string
---@param things E|E[]
local function do_something(things) end

do_something("foo")
do_something({"foo", "bar"}) -- this should match to `E[]` with `E=string` (or better yet with `E="foo"|"bar"`) and thus be valid, but instead tries to match the table as a whole to stringlib to satisfy `string` and warns that it doesn't satisfy the constraint
```

**Comment by justarandomgeek** [comment]
and, just for the point of comparison, in sumneko/luals, the library itself was typed as [`stringlib`](https://github.com/LuaLS/lua-language-server/blob/master/meta/template/string.lua) and then the value type was [`class string: stringlib`](https://github.com/LuaLS/lua-language-server/blob/master/meta/template/builtin.lua#L13), which seems like it would still allow for the extension you mention, while separating `string` from `stringlib` enough to make the table-or-string distinction properly
